src/lib/ilqp.js

```diff
@@ -0,0 +1,180 @@
+'use strict'
+
+const co = require('co')
+const Packet = require('./packet')
+const { startsWith, wait, xor, omitUndefined } = require('../utils')
+const debug = require('debug')('ilp:ilqp')
+const moment = require('moment')
+const BigNumber = require('bignumber.js')
+const uuid = require('uuid')
+
+function _sendAndReceiveMessage ({
+  plugin,
+  method,
+  message,
+  timeout
+}) {
+  const id = message.data.id = message.data.id || uuid()
+  debug('sending message:', JSON.stringify(message))
+
+  const responded = new Promise((resolve, reject) => {
+    function onIncomingMessage (response) {
+      debug('got incoming message:', JSON.stringify(response))
+      const data = response.data
+
+      if (!data || data.id !== id) return
+      if (data.method === 'error') reject(data.data.message)
+
+      if (data.method === method) {
+        debug('response of type', method)
+        plugin.removeListener('incoming_message', onIncomingMessage)
+        resolve(response)
+      }
+    }
+
+    // TODO: optimize to not add a listener each time?
+    plugin.on('incoming_message', onIncomingMessage)
+  })
+
+  return Promise.race([
+    plugin.sendMessage(message).then(() => responded),
+    wait(timeout || 5000)
+      .then(() => { throw new Error('quote request timed out') })
+  ])
+}
+
+function _getQuote ({
+  plugin,
+  connector,
+  quoteQuery,
+  timeout
+}) {
+  const prefix = plugin.getInfo().prefix
+
+  debug('remote quote connector=' + connector, 'query=' + JSON.stringify(quoteQuery))
+  return _sendAndReceiveMessage({
+    plugin: plugin,
+    method: 'quote_response',
+    timeout: timeout,
+    message: {
+      ledger: prefix,
+      account: connector,
+      data: {
+        method: 'quote_request',
+        data: quoteQuery
+      }
+    }
+  }).then((response) => {
+    return response.data.data
+  }).catch((err) => {
+    // TODO: should this do something else about errors?
```

Comment:
It's tricky because there are a couple of cases:

- multiple connectors, all return quotes - return the best quote
- multiple connectors, one throws an error while others return quotes - you want the quote and should ignore the errors
- multiple connectors, none respond - throw an error
- multiple connectors, multiple throw errors while none return quotes - ?